src/encoded/static/components/datatable.js

```diff
@@ -0,0 +1,182 @@
+ * );
+ * const tableData = {
+ *     rows: [
+ *         [1, 2, { content: 3, style: { fontWeight: 100 } }, 4, { content: <span style={{ fontWeight: 'bold' }}>5</span> }, 6],
+ *         [{ content: 7, css: 'cell-override' }, 8, { content: 'Special case', colSpan: 3 }, { content: contentFunc, value: 8 }],
+ *         { rowContent: [13, { content: () => <i>5</i> }, null, { content: contentFunc, value: 16 }, 17, 18], css: 'row-class' },
+ *         [19, 20, { content: contentFunc, value: 20, meta: { adder: 1, color: '#FFFF00' }}, 22, 23, 24]
+ *         [{ content: 'Full Width', colSpan: 0, style: { textAlign: 'center', backgroundColor: '#c0c0ff' } }],
+ *     ],
+ *     rowCss: 'overall-row',
+ *     tableCss: 'overall-table',
+ * };
+ *
+ * Displays:
+ * 1  2  3      4      5    6
+ * 7  8  Special case       3:1:8
+ * 13 5         3:2:16 17   18
+ * 19 20 2:3:21 22     23   24
+ *         Full Width
+ *
+ * ...where the "5" in the first row is bold as is the middle number in the two cells with
+ * `contentFunc` output, and the "3" appears lightweight. The first number in `contentFunc` output
+ * is italic, the middle is bold, and the last has a colored background. Each <tr> in the table has
+ * a CSS class of "overall-row" except the third which has "row-class". The "7" in the second row's
+ * <td> has a css class of "cell-override", and the "5" in the third row is italic. The third
+ * column of the third row is an empty <td> because of the null value in the `rowContent` array.
+ *
+ * Notes:
+ * - `rows` array is the only required property of `tableData`.
+ * - Each element of `row` must be either an array of elements to display in a single row, or an
+ *   object that must have a `rowContent` property.
+ * - The `rowContent` property must be an array of elements to display in a single row -- just like
+ *   a single element of the parent `rows` property.
+ *
+ * Why use `content` instead of a simple value?
+ *   - to add a CSS class or style to a specific cell
+ *   - to add a colSpan to a specific cell
+ *   - to have a cell rendered by a function returning jsx, or jsx itself
+ *
+ * Why use `rowContent` instead of an array of values and functions?
+ *   - to add a CSS class to the <tr> for a specific row
+ */
+
+import React from 'react';
+import PropTypes from 'prop-types';
+
+
+/**
+  * Determine the maximum number of columns of all rows in `DataTable` data.
+  * @param {object} tableData Same table data you pass to `DataTable`.
+  */
+const tableDataMaxWidth = (tableData) => {
+    const widths = tableData.rows.map(row => (Array.isArray(row) ? row.length : row.rowContent.length));
+    return Math.max(...widths);
+};
+
+
+/**
+ * Display a table of data with customizable styles per cell and per row, and with column spans.
+ */
+const DataTable = ({ tableData }) => {
+    let colNumber;
+    let cellContent;
+    let maxWidth;
```

Comment:
I did it this way because tableDataMaxWidth isn’t necessarily inexpensive if the table has a lot of columns, and we only need to call this if tableData has a 0 colSpan somewhere in it. If tableData has no colSpan:0 element, we never call tableDataMaxWidth.

Reply:
Ah yes, this makes sense.